### ISSUE TITLE
VB compiler emits bad IL

### ISSUE THREAD

**Issue by jrmoreno1** [dotnet/roslyn#9703]
**Version Used**: 
Version 14.0.24720.00 Update 1 of Visual Studio

**Steps to Reproduce**:

Build below in Release Mode
```
Public Class Form1
      Public Class BadCompiler
          Public Property Value As Date?
      End Class
      Private TestObj As BadCompiler = New BadCompiler()
      Public Sub IPE()
          Dim o as Object 
          o = TestObj.Value
      End Sub
  End Class
```

Run peverify against resulting exe

According to dotPeek, the IL is:

```
// [16 9 - 16 20]
IL_0000: ldarg.0      // this
IL_0001: ldfld        class WindowsApplication1.Form1/BadCompiler WindowsApplication1.Form1::t
IL_0006: callvirt     instance valuetype [mscorlib]System.Nullable`1<valuetype [mscorlib]System.DateTime> WindowsApplication1.Form1/BadCompiler::get_Value()
IL_000b: pop          
IL_000c: pop          

// [17 5 - 17 12]
IL_000d: ret          
```

**Expected Behavior**:
No errors reported.

**Actual Behavior**:
Stack underflow reported.  If the method is called, a InvalidProgramException occurs.

**Comment by jrmoreno1** [comment]
@gafter is the bug label only applied after it has been confirmed?

**Comment by jaredpar** [comment]
@jrmoreno1 sorry for the delay.  This bug got lost in the shuffle.  The bug label is added when it appears reasonable an issue could be a bug.  @gafter may have separately confirmed it but the application of the "bug" label doesn't necessarily mean it's true.  This will be looked at for RTM.

**Comment by TyOverby** [comment]
@jrmoreno1

Using your exact code with `vbc /target:library repro.vb` yields a dll that passes peverify.  

Using the following code:
```vb
Module MainModule
    Public Class Form1
        Public Class BadCompiler
            Public Property Value As Date?
        End Class

        Private TestObj As BadCompiler = New BadCompiler()

        Public Sub IPE()
            Dim o as Object 
            o = TestObj.Value
            Console.Write("value: ")
            Console.WriteLine(o)
        End Sub

    End Class

    Public Sub Main()
        Dim f = new Form1
        f.IPE()
    End Sub

End Module
```

I am able to compile the code (`vbc repro.vb`), it passes peverify, and then it runs without issue.  Since you filed this issue have you updated your compiler?  If so, could you verify that the issue is gone?  If so, I'll add this repro as a regression test and call it a day.

**Comment by jrmoreno1** [comment]
@TyOverby : I have upgraded, now using Version 14.0.25123.00 Update 2.  I see that I didn't mention that it only did this in release mode.  I'm having trouble getting PEVerify to work, but the app doesn't crash.  Although the IL as reported by dotNetPeek appears to be basically the same:

    .method public instance void 
       IPE() cil managed 
     {
        .maxstack 8

        // [8 9 - 8 26]
        IL_0000: ldarg.0      // this
        IL_0001: ldfld        class WindowsApplication1.Form1/BadCompilerWindowsApplication1.Form1::TestObj
        IL_0006: callvirt     instance valuetype [mscorlib]System.Nullable`1<valuetype [mscorlib]System.DateTime> WindowsApplication1.Form1/BadCompiler::get_Value()
        IL_000b: pop          
        IL_000c: pop          

        // [9 5 - 9 12]
        IL_000d: ret          

      } // end of method Form1::IPE

I'll look into it more soon.

**Comment by TyOverby** [comment]
@jrmoreno1 Ahh!  Compiling with `/optimize+` does cause this bug to surface.

Thanks!

**Comment by TyOverby** [comment]
Details: 

`EmitCallExpression` has a call to `EmitPopIfUnused`, and `EmitConversionExpression` has a call to `EmitPopIfUnused`.  This means that the code 
```vb
Dim o as Object = TestObj.Value
```
takes both of those paths and we wind up with an extra pop.

**Comment by jrmoreno1** [comment]
I apologize for leaving off the release mode, I would have sworn that I included it -- that's how it bit me, developed and tested in debug mode,  released and boom!
